hdf5 +fortran ^mpich %oneapi build fails during cmake: MakeTestFortranCompiler.cmake:61: ... unknown option '-rpath'

### Steps to reproduce the issue

`hdf5 ^mpich %oneapi` build fails with:
```
==> Installing hdf5-1.12.2-v62ghexbzbcmlrqginbiro4vuofqytc2
==> No binary for hdf5-1.12.2-v62ghexbzbcmlrqginbiro4vuofqytc2 found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/2a/2a89af03d56ce7502dcae18232c241281ad1773561ec00c0f0e8ee2463910f14.tar.gz
==> Ran patch() for hdf5
==> hdf5: Executing phase: 'cmake'
...
            MakeTestFortranCompiler.cmake:61 (message):
     331      The Fortran compiler
     332
     333        "/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-l26qdumkdmcfokuz6445pungqwz6yjoo/bin/mpif90"
     334
     335      is not able to compile a simple test program.
...
     348        ifx: command line warning #10006: ignoring unknown option '-rpath'
  >> 349    ld: cannot find /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-l26qdumkdmcfokuz6445pungqwz6yjoo/lib: file format not recog
            nized
...
```

Here using:
* `spack@develop` (5ffc095f63f5e6c82fe6289ae44066b492c37828 from `Thu Jul 21 17:02:28 2022 +0530`)
* Ubuntu 20.04, X86_64
* OneAPI 2022.1.0
* Docker container image `ecpe4s/ubuntu20.04-runner-x86_64-oneapi:2022-07-01`
  * contains OneAPI 2022.1.0 pre-installed to default location

Spack environment: [spack.yaml.txt](https://github.com/spack/spack/files/9160845/spack.yaml.txt)
```
spack:
  view: false

  concretizer:
    reuse: false
    unify: false

  compilers:
  - compiler:
      spec: oneapi@2022.1.0
      paths:
        cc: /opt/intel/oneapi/compiler/2022.1.0/linux/bin/icx
        cxx: /opt/intel/oneapi/compiler/2022.1.0/linux/bin/icpx
        f77: /opt/intel/oneapi/compiler/2022.1.0/linux/bin/ifx
        fc: /opt/intel/oneapi/compiler/2022.1.0/linux/bin/ifx
      flags: {}
      operating_system: ubuntu20.04
      target: x86_64
      modules: [compiler]
      environment: {}
      extra_rpaths: []

  packages:
    all:
      compiler: [oneapi]
      providers:
        blas: [openblas]
        mpi: [mpich]
      target: [x86_64]
      variants: +mpi

  specs:
  - hdf5@1.12.2 +fortran +hl +shared ^mpich~wrapperrpath
```

Concretization: [concretize.log](https://github.com/spack/spack/files/9160863/concretize.log)
```
==> Starting concretization
==> Environment concretized in 9.74 seconds.
==> Concretized hdf5@1.12.2+fortran+hl+shared ^mpich~wrapperrpath
 -   v62ghex  hdf5@1.12.2%oneapi@2022.1.0~cxx+fortran+hl~ipo~java+mpi+shared~szip~threadsafe+tools api=default build_type=RelWithDebInfo arch=linux-ubuntu20.04-x86_64
[+]  4522tfk      ^cmake@3.23.2%oneapi@2022.1.0~doc+ncurses+ownlibs~qt build_type=Release arch=linux-ubuntu20.04-x86_64
[+]  gvjgisz          ^ncurses@6.2%oneapi@2022.1.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-x86_64
[+]  euvcz47              ^pkgconf@1.8.0%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  rzfiqzd          ^openssl@1.1.1q%oneapi@2022.1.0~docs~shared certs=mozilla patches=3fdcf2d arch=linux-ubuntu20.04-x86_64
[+]  z4bgvwr              ^ca-certificates-mozilla@2022-03-29%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  v42mrbv              ^perl@5.34.1%oneapi@2022.1.0+cpanm+shared+threads arch=linux-ubuntu20.04-x86_64
[+]  ahtnphg                  ^berkeley-db@18.1.40%oneapi@2022.1.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu20.04-x86_64
[+]  44yzgyw                  ^bzip2@1.0.8%oneapi@2022.1.0~debug~pic+shared arch=linux-ubuntu20.04-x86_64
[+]  uli7ace                      ^diffutils@3.8%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  sohfg2f                          ^libiconv@1.16%oneapi@2022.1.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  mnowgf4                  ^gdbm@1.19%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  cd2lgld                      ^readline@8.1.2%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  5jz7ihy                  ^zlib@1.2.12%oneapi@2022.1.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-x86_64
[+]  l26qdum      ^mpich@4.0.2%oneapi@2022.1.0~argobots~cuda+fortran+hwloc+hydra+libxml2+pci~rocm+romio~slurm~two_level_namespace~vci~verbs~wrapperrpath datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu20.04-x86_64
[+]  rltl5ed          ^findutils@4.9.0%oneapi@2022.1.0 patches=440b954 arch=linux-ubuntu20.04-x86_64
[+]  ki4eyae          ^hwloc@2.7.1%oneapi@2022.1.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm+shared arch=linux-ubuntu20.04-x86_64
[+]  ky5qiaz              ^libpciaccess@0.16%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  5n46ly5                  ^libtool@2.4.7%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  agucw5d                      ^m4@1.4.19%oneapi@2022.1.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-x86_64
[+]  q56bku4                          ^libsigsegv@2.13%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  iwdrk44                  ^util-macros@1.19.3%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  wzm2qf3              ^libxml2@2.9.13%oneapi@2022.1.0~python arch=linux-ubuntu20.04-x86_64
[+]  bfylioq                  ^xz@5.2.5%oneapi@2022.1.0~pic libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  2k47srg          ^libfabric@1.14.1%oneapi@2022.1.0~debug~kdreg fabrics=sockets,tcp,udp arch=linux-ubuntu20.04-x86_64
[+]  t37lprz          ^yaksa@0.2%oneapi@2022.1.0~cuda~rocm arch=linux-ubuntu20.04-x86_64
[+]  svyjqmo              ^autoconf@2.69%oneapi@2022.1.0 patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-x86_64
[+]  lvuo5rb              ^automake@1.16.5%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  p7kwqnh              ^python@3.9.13%oneapi@2022.1.0+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-ubuntu20.04-x86_64
[+]  ijry5ro                  ^expat@2.4.8%oneapi@2022.1.0+libbsd arch=linux-ubuntu20.04-x86_64
[+]  flwxuwx                      ^libbsd@0.11.5%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  uxmtkav                          ^libmd@1.0.4%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  xoh5t2n                  ^gettext@0.21%oneapi@2022.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu20.04-x86_64
[+]  m2nt6nk                      ^tar@1.34%oneapi@2022.1.0 zip=pigz arch=linux-ubuntu20.04-x86_64
[+]  fisjzhz                          ^pigz@2.7%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  rg3ig4l                          ^zstd@1.5.2%oneapi@2022.1.0+programs compression=none libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  maq3kzw                  ^libffi@3.4.2%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]  7edvlef                  ^sqlite@3.38.5%oneapi@2022.1.0+column_metadata+dynamic_extensions+fts~functions+rtree arch=linux-ubuntu20.04-x86_64
[+]  k3c4qk3                  ^util-linux-uuid@2.37.4%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
```

Full logs below.


### Error message

<details><summary>Error message</summary><pre>
==> Installing hdf5-1.12.2-v62ghexbzbcmlrqginbiro4vuofqytc2
==> No binary for hdf5-1.12.2-v62ghexbzbcmlrqginbiro4vuofqytc2 found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/2a/2a89af03d56ce7502dcae18232c241281ad1773561ec00c0f0e8ee2463910f14.tar.gz
==> Ran patch() for hdf5
==> hdf5: Executing phase: 'cmake'
==> Error: ProcessError: Command exited with status 1:
    'cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/hdf5-1.12.2-v62ghexbzbcmlrqginbiro4vuofqytc2' '-DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo' '-DBUILD_TESTING:BOOL=OFF' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON' '-DCMAKE_INSTALL_RPATH:STRING=/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/hdf5-1.12.2-v62ghexbzbcmlrqginbiro4vuofqytc2/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/hdf5-1.12.2-v62ghexbzbcmlrqginbiro4vuofqytc2/lib64;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-l26qdumkdmcfokuz6445pungqwz6yjoo/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/hwloc-2.7.1-ki4eyae3mwpaiphpemqka3rypdiyzw2j/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/libpciaccess-0.16-ky5qiazdkxxis6hklky25q5rvusgnzus/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/libxml2-2.9.13-wzm2qf3umihsayypslqdvy3vrshagffl/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/libiconv-1.16-sohfg2fg2ss47uvlwzdjbifbjdbeg4wd/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/xz-5.2.5-bfylioquzdqx6woltxpqallazudls2eq/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/zlib-1.2.12-5jz7ihytig5qwohkhtyl6iqi3qppg4za/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/ncurses-6.2-gvjgiszsximd4475r2md2tnuyf2xzsab/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/libfabric-1.14.1-2k47srgm4fgdkw72nxvxfzcmcj5zeoah/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/yaksa-0.2-t37lprzrxkr4lowoep644r5oo6skxhq6/lib' '-DCMAKE_PREFIX_PATH:STRING=/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-l26qdumkdmcfokuz6445pungqwz6yjoo;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/yaksa-0.2-t37lprzrxkr4lowoep644r5oo6skxhq6;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/libfabric-1.14.1-2k47srgm4fgdkw72nxvxfzcmcj5zeoah;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/hwloc-2.7.1-ki4eyae3mwpaiphpemqka3rypdiyzw2j;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/libxml2-2.9.13-wzm2qf3umihsayypslqdvy3vrshagffl;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/xz-5.2.5-bfylioquzdqx6woltxpqallazudls2eq;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/libpciaccess-0.16-ky5qiazdkxxis6hklky25q5rvusgnzus;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/cmake-3.23.2-4522tfkc3hkawky76rjynn3nkgkz3u2x;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/zlib-1.2.12-5jz7ihytig5qwohkhtyl6iqi3qppg4za;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/libiconv-1.16-sohfg2fg2ss47uvlwzdjbifbjdbeg4wd;/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/ncurses-6.2-gvjgiszsximd4475r2md2tnuyf2xzsab' '-DALLOW_UNSUPPORTED:BOOL=ON' '-DHDF5_BUILD_EXAMPLES:BOOL=OFF' '-DBUILD_TESTING:BOOL=OFF' '-DHDF5_ENABLE_Z_LIB_SUPPORT:BOOL=ON' '-DHDF5_ENABLE_SZIP_SUPPORT:BOOL=OFF' '-DHDF5_ENABLE_SZIP_ENCODING:BOOL=OFF' '-DBUILD_SHARED_LIBS:BOOL=ON' '-DONLY_SHARED_LIBS:BOOL=OFF' '-DHDF5_ENABLE_PARALLEL:BOOL=ON' '-DHDF5_ENABLE_THREADSAFE:BOOL=OFF' '-DHDF5_BUILD_HL_LIB:BOOL=ON' '-DHDF5_BUILD_CPP_LIB:BOOL=OFF' '-DHDF5_BUILD_FORTRAN:BOOL=ON' '-DHDF5_BUILD_JAVA:BOOL=OFF' '-DHDF5_BUILD_TOOLS:BOOL=ON' '-DCMAKE_C_COMPILER:STRING=/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-l26qdumkdmcfokuz6445pungqwz6yjoo/bin/mpicc' '-DCMAKE_Fortran_COMPILER:STRING=/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-l26qdumkdmcfokuz6445pungqwz6yjoo/bin/mpif90' '/tmp/root/spack-stage/spack-stage-hdf5-1.12.2-v62ghexbzbcmlrqginbiro4vuofqytc2/spack-src'

4 errors found in build log:
     324    -- Found ZLIB: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/zlib-1.2.12-5jz7ihytig5qwohkhtyl6iqi3qppg4za/lib/libz.so (found version
            "1.2.12")
     325    -- The Fortran compiler identification is IntelLLVM 2022.1.0
     326    -- Detecting Fortran compiler ABI info
     327    -- Detecting Fortran compiler ABI info - failed
     328    -- Check for working Fortran compiler: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-l26qdumkdmcfokuz6445pungqwz6yjoo/bin
            /mpif90
     329    -- Check for working Fortran compiler: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-l26qdumkdmcfokuz6445pungqwz6yjoo/bin
            /mpif90 - broken
  >> 330    CMake Error at /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/cmake-3.23.2-4522tfkc3hkawky76rjynn3nkgkz3u2x/share/cmake-3.23/Modules/C
            MakeTestFortranCompiler.cmake:61 (message):
     331      The Fortran compiler
     332
     333        "/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-l26qdumkdmcfokuz6445pungqwz6yjoo/bin/mpif90"
     334
     335      is not able to compile a simple test program.
     336

     ...

     343        Building Fortran object CMakeFiles/cmTC_988c0.dir/testFortranCompiler.f.o
     344        /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-l26qdumkdmcfokuz6445pungqwz6yjoo/bin/mpif90   -fPIE -c /tmp/root/spack-
            stage/spack-stage-hdf5-1.12.2-v62ghexbzbcmlrqginbiro4vuofqytc2/spack-build-v62ghex/CMakeFiles/CMakeTmp/testFortranCompiler.f -o CMakeFiles/cmTC_9
            88c0.dir/testFortranCompiler.f.o
     345        Linking Fortran executable cmTC_988c0
     346        /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/cmake-3.23.2-4522tfkc3hkawky76rjynn3nkgkz3u2x/bin/cmake -E cmake_link_script CMakeF
            iles/cmTC_988c0.dir/link.txt --verbose=1
     347        /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-l26qdumkdmcfokuz6445pungqwz6yjoo/bin/mpif90 CMakeFiles/cmTC_988c0.dir/t
            estFortranCompiler.f.o -o cmTC_988c0
     348        ifx: command line warning #10006: ignoring unknown option '-rpath'
  >> 349    ld: cannot find /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-l26qdumkdmcfokuz6445pungqwz6yjoo/lib: file format not recog
            nized
  >> 350        make[1]: *** [CMakeFiles/cmTC_988c0.dir/build.make:99: cmTC_988c0] Error 1
     351        make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-hdf5-1.12.2-v62ghexbzbcmlrqginbiro4vuofqytc2/spack-build-v62ghex/CMakeFiles/CMa
            keTmp'
  >> 352        make: *** [Makefile:127: cmTC_988c0/fast] Error 2
     353
     354
     355
     356
     357
     358      CMake will not be able to correctly generate this project.

See build log for details:
  /tmp/root/spack-stage/spack-stage-hdf5-1.12.2-v62ghexbzbcmlrqginbiro4vuofqytc2/spack-build-out.txt
</pre></details>


### Information on your system

* **Spack:** 0.19.0.dev0 (5ffc095f63f5e6c82fe6289ae44066b492c37828)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/9160892/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/9160893/spack-build-out.txt)
[CMakeError.log](https://github.com/spack/spack/files/9160900/CMakeError.log)
[CMakeOutput.log](https://github.com/spack/spack/files/9160901/CMakeOutput.log)

@ChristopherHogan @brtnfld @byrnHDF @epourmal @gheber @hyoklee @lkurz @lrknox @soumagne @rscohn2 @raffenet @yfguo @wspear

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/mpich/mpich-oneapi-config-rpath.patch
+++ b/var/spack/repos/builtin/packages/mpich/mpich-oneapi-config-rpath.patch
@@ -1,0 +1,30 @@
+diff -ruN spack-src/confdb/config.rpath spack-src-patched/confdb/config.rpath
+--- spack-src/confdb/config.rpath	2022-03-29 15:13:49.000000000 -0700
++++ spack-src-patched/confdb/config.rpath	2022-07-25 17:54:14.638367460 -0700
+@@ -76,7 +76,7 @@
+         ecc*)
+           wl='-Wl,'
+           ;;
+-        icc* | ifort*)
++        icc* | icx* | ifort* | ifx*)
+           wl='-Wl,'
+           ;;
+         lf95*)
+@@ -233,7 +233,7 @@
+           enable_dtags_flag="${wl}--enable-new-dtags"
+           disable_dtags_flag="${wl}--disable-new-dtags"
+         else
+-          case $cc_basename in ifort*)
++          case $cc_basename in ifort* | ifx*)
+             enable_dtags_flag="${wl}--enable-new-dtags"
+             disable_dtags_flag="${wl}--disable-new-dtags"
+             ;;
+@@ -380,7 +380,7 @@
+       ;;
+     darwin* | rhapsody*)
+       hardcode_direct=no
+-      if { case $cc_basename in ifort*) true;; *) test "$GCC" = yes;; esac; }; then
++      if { case $cc_basename in ifort* | ifx*) true;; *) test "$GCC" = yes;; esac; }; then
+         :
+       else
+         ld_shlibs=no

--- a/var/spack/repos/builtin/packages/mpich/package.py
+++ b/var/spack/repos/builtin/packages/mpich/package.py
@@ -140,6 +140,9 @@ with '-Wl,-commons,use_dylibs' and without
         'mpicc', 'mpicxx', 'mpif77', 'mpif90', 'mpifort', relative_root='bin'
     )
 
+    # https://github.com/spack/spack/issues/31678
+    patch('mpich-oneapi-config-rpath.patch', when='@4.0.2 %oneapi')
+
     # Fix using an external hwloc
     # See https://github.com/pmodels/mpich/issues/4038
     # and https://github.com/pmodels/mpich/pull/3540

